### PR TITLE
feat: assign terms to CustomEvent

### LINKED ISSUE #762: Download ICS: Assign CourseEvent term to CustomEvent
Right now we are not assigning specific terms to a custom event when a user creates it.

We will grab the term from CourseEvent in the schedule.
- if the terms are inconsistent (e.g. some classes are in the 2023 Fall, some are in the 2023 Spring), pick the latest one.
- if there is no CourseEvent in the schedule, use the default term.

### PR DESCRIPTION
**Pull request by training4usaco**
## Context
- This PR is picking up the work done by @Slo1k in #1098 
- In short, currently, custom events do not have any term associated with them. 
## Summary
- After this PR, whenever a user creates a custom event, we will automatically assign a term to the custom event
- The term is determined by the courses in the schedules provided for the custom event as follows:
    - If there are no course events, `defaultTerm` will be picked
    - Otherwise, the latest term out of the course events will be picked
- The term is then stored in the database in a column called "term" in the table `customEvent`
- Updated both customEvent and courseEvent UI to use MUI components
- Note 1: deleting/adding regular course events will not update custom event terms, unless the custom event is edited and saved (there is also no option to pick the term for the custom event manually)
- Note 2: I'm not sure if the snapshot should get merged as well, but it accounts for +450, so the PR itself is around +150 -41

## Test Plan
1. Test creation of a custom event with a single term in the schedule. Verify that the custom event is assigned the correct term.
2. Test creation of a custom event with multiple terms in the schedule. Ensure the latest term is assigned to the custom event.
3. Test creation of a custom event with no associated terms in the schedule. Verify that the default term is applied.
4. Ensure that edge cases like missing terms or incorrect term names in the schedule are handled correctly and fall back to the default term.
5. Verify that custom events are correctly displayed with their respective terms in calendar view - CustomEventDialog.
6. Test loading and saving to backend. Verify that loaded and saved custom events have the appropriate term.

## Issues

Closes https://github.com/icssc/AntAlmanac/issues/762

### CODE REVIEW

**Comment by training4usaco** [comment]
@alexespejo Hey! Could I get a review for this PR. It's pretty similar to the linked PR. I also left a couple notes in the PR comment about things I wasn't particularly sure about.

File: apps/antalmanac/src/components/Calendar/CourseCalendarEvent.tsx [hunk 1]
```diff
@@ -231,16 +231,28 @@ export const CourseCalendarEvent = ({ selectedEvent, scheduleNames, closePopover
...
         return (
             <Paper sx={{ padding: '0.5rem' }} ref={paperRef}>
                 <Box sx={{ fontSize: '0.9rem', fontWeight: 500 }}>{title}</Box>
-                {building && (
-                    <Box sx={{ border: 'none', width: '100%', borderCollapse: 'collapse', fontSize: '0.9rem' }}>
-                        Location:&nbsp;
-                        <MapLink buildingId={+building} room={buildingCatalogue[+building]?.name ?? ''} />
-                    </Box>
-                )}
+                <table style={{ border: 'none', width: '100%', borderCollapse: 'collapse', fontSize: '0.9rem' }}>
```
Comment (alexespejo) [inline apps/antalmanac/src/components/Calendar/CourseCalendarEvent.tsx]: issue: we use a library called MUI that that provides dedicated components such as `<Table>`

Comment (training4usaco) [reply apps/antalmanac/src/components/Calendar/CourseCalendarEvent.tsx]: Updated!

File: apps/antalmanac/src/components/Calendar/CourseCalendarEvent.tsx [hunk 2]
```diff
@@ -208,39 +233,62 @@ export const CourseCalendarEvent = ({ selectedEvent, scheduleNames, closePopover
...
-                            <td>Color</td>
-                            <td style={{ textAlign: 'right' }}>
+                            </TableCell>
+                        </TableRow>
+                        <TableRow>
+                            <TableCell>Final</TableCell>
+                            <TableCell sx={{ textAlign: 'right' }}>{finalExamString}</TableCell>
+                        </TableRow>
+                        <TableRow>
+                            <TableCell style={{ paddingBottom: 0 }}>Color</TableCell>
```
Comment (alexespejo) [inline apps/antalmanac/src/components/Calendar/CourseCalendarEvent.tsx:243]: nit: sx prop

File: apps/antalmanac/src/components/Calendar/CourseCalendarEvent.tsx [hunk 3]
```diff
@@ -208,39 +233,62 @@ export const CourseCalendarEvent = ({ selectedEvent, scheduleNames, closePopover
...
+                            <TableCell style={{ textAlign: 'right', paddingBottom: 0 }}>
```
Comment (alexespejo) [inline apps/antalmanac/src/components/Calendar/CourseCalendarEvent.tsx:244]: nit: use sx prop

File: apps/antalmanac/src/components/Calendar/CourseCalendarEvent.tsx [hunk 4]
```diff
@@ -208,39 +233,62 @@ export const CourseCalendarEvent = ({ selectedEvent, scheduleNames, closePopover
...
+                        },
+                        border: 'none',
+                        width: '100%',
+                        borderCollapse: 'collapse',
+                        fontSize: '0.9rem',
+                    }}
+                >
+                    <TableBody>
+                        <TableRow>
+                            <TableCell style={{ paddingTop: '0.2rem', paddingRight: '2rem' }}>Term</TableCell>
```
Comment (alexespejo) [inline apps/antalmanac/src/components/Calendar/CourseCalendarEvent.tsx:279]: nit: use sx prop (sorry if this starts getting redundant)

File: apps/antalmanac/src/components/Calendar/CourseCalendarEvent.tsx [hunk 5]
```diff
@@ -208,39 +233,62 @@ export const CourseCalendarEvent = ({ selectedEvent, scheduleNames, closePopover
...
+                            <TableCell sx={{ textAlign: 'right' }}>{term}</TableCell>
+                        </TableRow>
+                        {building && (
+                            <TableRow>
+                                <TableCell style={{ paddingRight: '2rem' }}>Location</TableCell>
```
Comment (alexespejo) [inline apps/antalmanac/src/components/Calendar/CourseCalendarEvent.tsx:284]: nit: use sx prop

**Review by alexespejo — CHANGES_REQUESTED** [review]
Please resolve the merge conflict. Thanks!
